[16.0] [ADD] website_slides_attendees_completed_time: show completed time in attendee views

rrebollo opened the pull request:
This module extends the functionality of course attendee views to show course's completed time.

After module installation, you will see course's completed time within attendee views.

![attendees_tree_view](https://github.com/user-attachments/assets/22454d9a-485e-4cb5-a7ce-9f6ab168aeba)

@BinhexTeam

rrebollo committed:
[ADD] website_slides_attendees_completed_time: implement completed time tracking in attendee views

[IMP] website_slides_attendees_completed_time: compute attendee completed time through built-in recompute_completion feature

[IMP] website_slides_attendees_completed_time: enhance DESCRIPTION; implement tests for completed time

rrebollo commented:
![image](https://github.com/user-attachments/assets/b5f6f23a-a4ac-481b-ad49-349b803fd2ae)

christian-ramos-tecnativa commented on website_slides_attendees_completed_time/models/slide_channel_partner.py:
```diff
@@ -0,0 +1,34 @@
+        "data_slide_partner_ids.completed",
+        "data_slide_partner_ids.slide_id.completion_time",
+    )
+    def _compute_completed_time(self):
```
In order to avoid creating this compute you could extend the function `_recompute_completion` https://github.com/odoo/odoo/blob/40f1b3ffc8c611dc80cb9b806fdb28baf8df2c2c/addons/website_slides/models/slide_channel.py#L49 and assign the value to `completed_time` after `super()`. This way you dont even need the variable `data_slide_partner_ids` as you could get the information by `search()`

rrebollo replied:
During the early development phase, I explored both options:

1. `_recompute_completion` approach:
   - Briefly investigated but didn't dive deep
   - Encountered synchronization issues during manual testing with demo data
   - Possible causes:
     - My own limited functional expertise with this method
     - Potential timing/triggering complexities

2. Compute method approach:
   - Ultimately chose this path
   - More predictable triggering behavior
   - Felt more certain about recomputation control

**Next Steps:**
- Keeping current commit unsquashed for reference during experiments
- Will reconsider the `_recompute_completion` approach to:
  - Better understand built-in philosophy
  - Evaluate if it could provide cleaner implementation
  - Verify if synchronization issues were environment-specific

christian-ramos-tecnativa commented on website_slides_attendees_completed_time/models/slide_slide_partner.py:
```diff
@@ -0,0 +1,27 @@
+    )
+
+    @api.depends("channel_id", "partner_id")
+    def _compute_channel_partner_id(self):
```
With the approximation above you shouldn't need this file

christian-ramos-tecnativa reviewed (changes requested):
Let me know if I explained myself properly. Thanks

christian-ramos-tecnativa reviewed (approved):
Functional and code review, LGTM

rrebollo commented:
## ✅ Update on the Approach Taken

Hello @Christian-RB and community,

After careful consideration, we’ve decided to align with the native **`_recompute_completion`** approach for tracking course progress. This decision has greatly simplified the logic and integration, and the overall flow of the feature now feels much smoother.

### 🔄 What Changed

The key update alongside this decision is that the field **`completed_time`** is **no longer a computed field**. Instead, it’s now a **stored field** that must be initialized with data when the addon is installed.

To ensure this, we’ve implemented a **`post_init_hook`**, which populates the `completed_time` values for all existing attendee records based on their current completed slides.

### 🧪 Tests Included

We’ve also added a suite of unit tests. These were designed following the pattern of the native tests for the **`completion`** field (which corresponds to the **Progress** column in the attendee view). 

It’s important to highlight that although the `completed_time` and `completion` fields share similar purposes, their logic is **not interchangeable**:
- `completion` (Progress) is **slide-count based**
- `completed_time` is **time-duration based** and must remain an **accurate, real-time reflection** of the attendee’s completed course content

This means that updates to `completed_time` must respond to changes like:
- Slides being archived or restored
- Slides being marked as incomplete
- Course structure being modified

We believe this dual-tracking (count-based + time-based) gives a more robust and insightful representation of learner engagement.

---

We’re now just awaiting the community’s feedback. Thanks for reviewing!

christian-ramos-tecnativa commented:
@rrebollo Can you add the version [16.0] in the PR title? Thanks

edescalona reviewed (approved):
LGTM

rrebollo commented:
- @jelenapoblet maybe we could reopen this in order to get it merged?

jelenapoblet commented:
Hi @rrebollo is this necessary when a progress field already exists?

rrebollo commented:
@jelenapoblet we anticipate this concern. From a comment above 👉:
> It’s important to highlight that although the `completed_time` and `completion` fields share similar purposes, their logic is **not interchangeable**:
> - `completion` (Progress) is **slide-count based**
> - `completed_time` is **time-duration based** and must remain an **accurate, real-time reflection** of the attendee’s completed course content

And from the addon's documentation:
>**Built-in (Odoo):** 
> - ``Progress = Number of Completed Slides / Total Number of Slides`` → 
    A simple ratio, based only on slide count. 
 >**Added by this addon:** 
> - ``Completed Time = Σ(completion_time of each completed slide)`` → A 
    cumulative total of time-based effort for each attendee. 
 
>🧪 Example 
 >^^^^^^^^^^
> For a course with 10 slides (total 20 hours), if an attendee completes: 
> -  1 slide of 15 hours → **Progress**: 10% → **Completed Time**: 15h

rrebollo commented:
ping @jelenapoblet [comment](https://github.com/OCA/e-learning/pull/10#issuecomment-3798967155)

jelenapoblet commented:
@rrebollo are you also going to maintain this ? is there a plan to migrate this module to v17, 18 and 19 ?

rrebollo commented:
> @rrebollo are you also going to maintain this ? is there a plan to migrate this module to v17, 18 and 19 ?

Yes and yes

jelenapoblet commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump
